ext/random/randomizer.c

```diff
@@ -0,0 +1,304 @@
+	zend_object *engine_object = NULL;
+	zval zengine_object;
+	zend_string *pname;
+
+	ZEND_PARSE_PARAMETERS_START(0, 1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS_OR_NULL(engine_object, random_ce_Random_Engine);
+	ZEND_PARSE_PARAMETERS_END();
+
+	/* Create default RNG instance */
+	if (!engine_object) {
+		engine_object = random_ce_Random_Engine_Secure->create_object(random_ce_Random_Engine_Secure);
+
+		/* No need self-refcount */
+		GC_DELREF(engine_object);
+	}
+
+	ZVAL_OBJ(&zengine_object, engine_object);
+
+	/* Write property */
+	pname = zend_string_init("engine", sizeof("engine") - 1, 0);
+	zend_std_write_property(&randomizer->std, pname, &zengine_object, NULL);
+	zend_string_release(pname);
+
+	randomizer_common_init(randomizer, engine_object);
+}
+/* }}} */
+
+/* {{{ Generate random number in range */
+PHP_METHOD(Random_Randomizer, getInt)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	uint64_t result;
+	zend_long min, max;
+	int argc = ZEND_NUM_ARGS();
+
+	if (argc == 0) {
+		result = randomizer->algo->generate(randomizer->status);
+		if (randomizer->status->last_generated_size > sizeof(zend_long)) {
+			zend_throw_exception(spl_ce_RuntimeException, "Generated value exceeds size of int", 0);
+			RETURN_THROWS();
+		}
+		if (randomizer->status->last_unsafe) {
+			zend_throw_exception(spl_ce_RuntimeException, "Random number generation failed", 0);
+			RETURN_THROWS();
+		}
+		RETURN_LONG((zend_long) (result >> 1));
+	}
+
+	ZEND_PARSE_PARAMETERS_START(2, 2)
+		Z_PARAM_LONG(min)
+		Z_PARAM_LONG(max)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (UNEXPECTED(max < min)) {
+		zend_argument_value_error(2, "must be greater than or equal to argument #1 ($min)");
+		RETURN_THROWS();
+	}
+
+	result = randomizer->algo->range(randomizer->status, min, max);
+	if (randomizer->status->last_unsafe) {
+		zend_throw_exception(spl_ce_RuntimeException, "Random number generation failed", 0);
+		RETURN_THROWS();
+	}
+
+	RETURN_LONG((zend_long) result);
+}
+/* }}} */
+
+/* {{{ Generate random bytes string in ordered length */
+PHP_METHOD(Random_Randomizer, getBytes)
+{
+	php_random_randomizer *randomizer = Z_RANDOM_RANDOMIZER_P(ZEND_THIS);
+	zend_string *retval;
+	zend_long length;
+	uint64_t result;
+	size_t total_size = 0, required_size;
+	uint32_t i;
+
+	ZEND_PARSE_PARAMETERS_START(1, 1)
+		Z_PARAM_LONG(length)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (length < 1) {
+		zend_argument_value_error(1, "must be greater than 0");
+		RETURN_THROWS();
+	}
+
+	retval = zend_string_alloc(length, 0);
+	required_size = length;
+
+	while (total_size < required_size) {
+		result = randomizer->algo->generate(randomizer->status);
+		if (randomizer->status->last_unsafe) {
+			zend_string_free(retval);
+			zend_throw_exception(spl_ce_RuntimeException, "Random number generation failed", 0);
+			RETURN_THROWS();			
+		}
+		for (i = 0; i < randomizer->status->last_generated_size; i++) {
+			ZSTR_VAL(retval)[total_size++] = (result >> (i * 8) & 0xff);
```

Comment:
Nit: did you mean to parenthesize this?

```suggestion
			ZSTR_VAL(retval)[total_size++] = (result >> (i * 8)) & 0xff;
```

Reply:
Exactly! Thank you.